src/shogun/distance/EuclideanDistance.h

```diff
@@ -125,6 +161,21 @@ class CEuclideanDistance: public CRealDistance
 	protected:
 		/** if application of sqrt on matrix computation is disabled */
 		bool disable_sqrt;
+
+		/** if squared norms of rhs features are used */
+		bool m_rhs_sq_norms;
+
+		/** if squared norms of lhs features are used */
+		bool m_lhs_sq_norms;
+
+		/** i dot products are used in distance computations */
+		bool m_dot_enabled;
+
+		/** squared norms from features of right hand side to be used if dot enabled */
+		SGVector<float64_t> m_rhs_squared_norms;
```

Comment:
Actually I had another thought about this.

It should be like this (and this is close to what you did initially)

By default, nothing is precomputed (we dont know which distances user wants to access)
 and both the rhs and lhs norms are just empty vectors
There is an optional method "precompute" (plus two method precompute_lhs and precompute_rhs that precomputes both sides
Whenever rhs or lhs are changed, the norm vectors are reset.

There is a method to compute the distance for *all* pairs of lhs/rhs. This one obviously calls precompute inside

Then you can manually call the precompute in KMeans and/or ask for many distances